### PR TITLE
Add support for setting storage size on ZFS containers

### PR DESCRIPTION
**Pull request by kenXengineering**
**\- What I did**
Added support for for the `--storage-opt` CLI flag to ZFS.  Users can specify a container's block device size by passing size (`--storage-opt size=2G`).  Built off of pull request https://github.com/docker/docker/pull/19367

**\- How I did it**
When creating or cloning a ZFS volume, check storage options for the key `size`.  If it is provided, set the ZFS property `quota` to the given value on the new ZFS filesystem.  If size is not specified, do not set `quota`.

**\- How to verify it**
Have docker engine setup with ZFS as the storage driver.  Pull an image and run it with the `--storage-opts size=<size>` flag.  Run `df -h` to view attached volumes sizes.  The `Avail` space on the root device should be the size specified.

```
root@dockerzfs:~# docker run -it --rm --storage-opt size=2G ubuntu bash
root@217c5a71c117:/# df -h
Filesystem                                                                            Size  Used Avail Use% Mounted on
zpool-docker/docker/a9897bab72a60873898191704eb0dfff96aa4efdb0a2676d9e8a74ad20b265a7  2.2G  198M  2.0G   9% /
tmpfs                                                                                1001M     0 1001M   0% /dev
tmpfs                                                                                1001M     0 1001M   0% /sys/fs/cgroup
zpool-docker/docker                                                                    48G  3.2M   48G   1% /etc/hosts
shm                                                                                    64M     0   64M   0% /dev/shm
```

Note:  Due to the way docker utilizes ZFS, the newly created ZFS volume will be from a read-only snapshot of the docker image.  Therefore setting the `quota` on the new volume will make the available space the size of the quota as it is applied to the new volume and not parent volumes.

**\- A picture of a cute animal**
Westies at the Cincinnati Reds Parade

This is my first pull request to Docker, and I welcome all comments and suggestions!

Signed-off-by: Ken Herner kherner@progress.com

### CODE REVIEW

**Comment by calavera** [comment]
Nice! This was in my list of improvements for ZFS.

**Comment by thaJeztah** [comment]
experimental is green; https://jenkins.dockerproject.org/job/Docker-PRs-experimental/17500/console

```
22:17:40 Archiving artifacts
22:17:41 Notifying endpoint 'HTTP:https://leeroy.dockerproject.org/notification/jenkins'
22:17:41 Finished: SUCCESS
```

**Comment by calavera** [comment]
This would be easier to implement after merging https://github.com/docker/docker/pull/21139.

**Comment by calavera** [comment]
we should have global quota(at the daemon level) and local quota(what this PR implements).

**Comment by chosenken** [comment]
@calavera @thaJeztah Hello.  I haven't seen any movement on the pull request.  Do you have any questions or comments?

**Comment by calavera** [comment]
Design LGTM. I moved it to code-review so others can give feedback.

File: daemon/graphdriver/zfs/zfs.go [hunk 1]
```diff
@@ -271,22 +267,55 @@ func (d *Driver) Create(id string, parent string, mountLabel string, storageOpt
...
 
 	// retry
-	return d.create(id, parent)
+	return d.create(id, parent, storageOpt)
 }
 
-func (d *Driver) create(id, parent string) error {
+func (d *Driver) create(id, parent string, storageOpt map[string]string) error {
 	name := d.zfsPath(id)
+	quota, err := parseStorageOpt(storageOpt)
```
Comment (thaJeztah) [inline daemon/graphdriver/zfs/zfs.go]: Should `err` be checked here?

Comment (thaJeztah) [inline daemon/graphdriver/zfs/zfs.go]: Perhaps return early on error (also below), because there's now a lot of nested if's (but that's just a suggestion)

**Comment by cpuguy83** [comment]
Looks ok to me, but I don't use zfs.

**Comment by thaJeztah** [comment]
ping @chosenken could you address my comment? Also don't forget to update the relevant documentation and man-pages, but let me know if you need hints for which ones need an update

**Comment by Mic92** [comment]
For reference I also made a pull request for quotas for zfs in the past: https://github.com/docker/docker/pull/12520

**Comment by chosenken** [comment]
@thaJeztah I've updated that section to return on error, thanks for the review!  As for documentation, I think the details added in this commit https://github.com/docker/docker/commit/b16decfccfdb0749c490be9272cb7b4789be87b4#diff-505c72218d90da970c16fdbf0b4f613c should cover it.  I'm using the same option from #19367.  I'm also going to add tests for this, so I will update when I get them added.

File: daemon/graphdriver/zfs/zfs.go [hunk 2]
```diff
@@ -273,22 +269,58 @@ func (d *Driver) Create(id string, parent string, mountLabel string, storageOpt
...
 	}
-	return d.cloneFilesystem(name, d.zfsPath(parent))
+	err = d.cloneFilesystem(name, d.zfsPath(parent))
+	if err == nil {
+		err = setQuota(name, quota)
+	}
+	return err
+}
+
+func parseStorageOpt(storageOpt map[string]string) (string, error) {
```
Comment (calavera) [inline daemon/graphdriver/zfs/zfs.go]: You could just replace this function with:

```go
func parseStorageOpt(storageOpt map[string]string) (string) {
    return storageOpt["size"]
}
```

**Comment by chosenken** [comment]
I've added a test case based off of @Mic92 pull request #12520.

**Comment by calavera** [comment]
LGTM.

It looks like there is a flaky test:

```
19:26:02 FAIL: docker_cli_run_unix_test.go:947: DockerSuite.TestRunSeccompDefaultProfile
19:26:02 
19:26:02 docker_cli_run_unix_test.go:1045:
19:26:02     c.Assert(err, checker.IsNil)
19:26:02 ... value *errors.errorString = &errors.errorString{s:"expected No such file or directory, got: /go/src/github.com/docker/docker/bundles/1.12.0-dev/test-integration-cli/../binary-client/docker: Error response from daemon: rpc error: code = 2 desc = \"containerd: container did not start before the specified timeout\".\n"} ("expected No such file or directory, got: /go/src/github.com/docker/docker/bundles/1.12.0-dev/test-integration-cli/../binary-client/docker: Error response from daemon: rpc error: code = 2 desc = \"containerd: container did not start before the specified timeout\".\n")
```

**Comment by cpuguy83** [comment]
LGTM

**Comment by barrachri** [comment]
Hello, does this work only with containers?

Because if I follow the docs from here: https://docs.docker.com/engine/userguide/storagedriver/zfs-driver/#configure-docker-with-the-zfs-storage-driver

I just get this error https://github.com/moby/moby/issues/33847

**Pull request merged**